nrf_security/cmake/symbol_rename.cmake

```diff
@@ -12,50 +12,108 @@
 # If two or more backends are enabled for a given function, each of them will
 # be renamed to be unique and a glue layer will be used as new front-end.
 #
+
+#
+# This macro comments out the instruction to rename a symbol
+# in case the configuration is not glued and enabled in backend
+#
 macro(keep_config_test_glue mbedtls_config backend_name)
   if(NOT CONFIG_GLUE_${mbedtls_config}
      AND CONFIG_${backend_name}_${mbedtls_config})
     set(KEEP_${mbedtls_config} "#")
   endif()
 endmacro()
 
+#
+# This macro comments out the instruction to strip a symbol
+# if the configuration is not enabled in the backend
+#
 macro(keep_config_test mbedtls_config backend_name)
   if(CONFIG_${backend_name}_${mbedtls_config})
     set(KEEP_${mbedtls_config} "#")
   endif()
 endmacro()
 
-function(library_redefine_symbols backend template)
+#
+# This macro will create a string to remove objects
+# if the configuration is not enabled in the backend
+#
+macro(remove_objects mbedtls_config backend_name object_list object_name)
+  if (NOT CONFIG_${backend_name}_${mbedtls_config})
+    kconfig_mbedtls_debug("Removing object ${backend_name}:${object_name}")
+    set(${object_list} ${${object_list}} ${object_name})
+  endif()
+endmacro()
+
+function(symbol_rename_func backend rename_template is_stripped)
   string(TOUPPER "${backend}" BACKEND_NAME)
-  keep_config_test_glue("MBEDTLS_AES_C" ${BACKEND_NAME})
+  #
+  # Files that can be glued
+  #
+  keep_config_test_glue("MBEDTLS_AES_C"           ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_CIPHER_MODE_XTS" ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_CIPHER_MODE_CBC" ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_CIPHER_MODE_CFB" ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_CIPHER_MODE_OFB" ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_CIPHER_MODE_CTR" ${BACKEND_NAME})
-  keep_config_test("MBEDTLS_SHA1_C"               ${BACKEND_NAME})
-  keep_config_test("MBEDTLS_SHA256_C"             ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_RSA_C"           ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_DHM_C"           ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_CCM_C"           ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_CMAC_C"          ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_ECDH_C"          ${BACKEND_NAME})
   keep_config_test_glue("MBEDTLS_ECDSA_C"         ${BACKEND_NAME})
-  keep_config_test_glue("MBEDTLS_GCM_C"           ${BACKEND_NAME})
-  keep_config_test("MBEDTLS_PLATFORM"             ${BACKEND_NAME})
-  
+
+  #
+  # Files that are either in one of the backends
+  # These will be removed with a call to "ar"
```

Comment:
Why do we need to remove object files from the archive?

Usually, if you don't want something from an archive you will make sure there are no references to it and the linker will then not include it.

Reply:
Also,

should be

Files that are in either one of the backends

Reply:
We remove objects from the archive in case we replace the functionality by a different backen (e.g. vanilla). The binary library in question (CC310 backend) will contain symbols we don't want in this case...

Will fix the text

Reply:
We remove object files from the archive as the functionality is colliding with a replacement in a different backend (e.g. implemented in SW vanilla instead of HW accelerated from pre-compiled CC310 library)

Reply:
When module A and B offer the same symbol and A has priority then this should be resolved by placing A before B in the link command.

Reply:
We don't wan't colliding symbols. We want to be absolutely sure we have the correct symbol as the context-sizes are different and the code will still run with undefined behavior (read: crashing).

Won't fix for security reasons